hal/src/common/dmac/channel.rs

```diff
@@ -0,0 +1,357 @@
+//! # Abstractions over individual DMA channels
+//!
+//! # Initializing
+//!
+//! Individual channels should be initialized through the
+//! [`Channel::init`] method. This will return a `Channel<Id, Ready>` ready for
+//! use by a [`Transfer`](super::transfer::Transfer). Initializing a channel
+//! requires setting a priority level, as well as enabling or disabling
+//! interrupt requests (only for the specific channel being initialized).
+//!
+//! # Burst Length and FIFO Threshold (SAMD51/SAME5x only)
+//!
+//! The transfer burst length can be configured through the
+//! [`Channel::burst_length`] method. A burst is an atomic,
+//! uninterruptible transfer which length corresponds to a number of beats. See
+//! SAMD5x/E5x datasheet section 22.6.1.1 for more information. The FIFO
+//! threshold can be configured through the
+//! [`Channel::fifo_threshold`] method. This enables the channel
+//! to wait for multiple Beats before sending a Burst. See SAMD5x/E5x datasheet
+//! section 22.6.2.8 for more information.
+//!
+//! # Channel status
+//!
+//! Channels can be in any of three statuses: [`Uninitialized`], [`Ready`], and
+//! [`Busy`]. These statuses are checked at compile time to ensure they are
+//! properly initialized before launching DMA transfers.
+//!
+//! # Resetting
+//!
+//! Calling the [`Channel::reset`] method will reset the channel to its
+//! `Uninitialized` state. You will be required to call [`Channel::init`]
+//! again before being able to use it with a `Transfer`.
+
+use super::dma_controller::{ChId, DmaController, PriorityLevel, TriggerAction, TriggerSource};
+use crate::{
+    target_device::DMAC,
+    typelevel::{Is, Sealed},
+};
+use core::{marker::PhantomData, mem};
+
+#[cfg(feature = "min-samd51g")]
+use super::dma_controller::{BurstLength, FifoThreshold};
+
+#[cfg(feature = "min-samd51g")]
+use crate::target_device::dmac::CHANNEL;
+
+//==============================================================================
+// Channel Status
+//==============================================================================
+pub trait Status: Sealed {}
+
+/// Uninitialized channel
+pub struct Uninitialized;
```

Comment:
This is 100% a personal preference thing, but you can also implement these types as `enum`s. There are two main differences, as I see it. First, as `struct`s, you can instantiate them. But when they are blank `enum`s, they can't actually be instantiated at all. Rather, they only exist at the type level. To instantiate them, you need `PhantomData`. And second, they are grouped with the other `enum`s in the documentation. I prefer both of those options, so I typically opt for declaring them as `enum`s. I also view these as type-level variants of a type-level `enum` (the `Status` trait), so I prefer to group them with `enum`s.

Just thought I would add some food for thought.